ROX-17011: Refactor compliance to enable local-compliance

vikin91 opened the pull request:
## Description

This PR extracts all runtime dependencies of compliance into interfaces (refactoring part), so that compliance can now easily change behavior of:
1. ~Logging~
2. Generating Node-Inventories
3. Getting the name of the node it runs on
4. Handling the ACK/NACK messages from Sensor

The behavior of the Compliance component has not been changed in this PR.

This PR adds `local-compliance` that can be run locally and connect to the real Sensor (more info about this in the readme). The refactoring part was necessary to for `local-compliance` to happen. 

This PR is a prerequisite to enable Sensor integration tests in the scope defined in [ROX-16861](https://issues.redhat.com/browse/ROX-16861) 

## Checklist
- [x] Investigated and inspected CI test results


### N/A
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

## Testing Performed

- Run local-compliance as documented in the readme
- Change compliance image to image from this branch on running cluster and confirm that it still runs as expected.



Based on a work started in branch `jschnath/ROX-17011_Investigate_the_viability_of_a_local_compliance`. Once merged, this branch can be deleted.

vikin91 commented:
/test gke-qa-e2e-tests

vikin91 commented:
/retest

fredrb commented on tools/local-compliance/loadGeneratingNodeScanner.go:
```diff
@@ -0,0 +1,64 @@
+// LoadGeneratingNodeScanner is a scanner that generates fake scans with high frequecy of the node-inventory messages.
+// Its main purpose is to generate load for load-testing of Sensor
+type LoadGeneratingNodeScanner struct {
+	log          *logging.Logger
```
Why are you providing a logger as a dependency? 

I think the common pattern in the project is to define the log locally in each package like:
```
var (
     log = logging.LoggerForModule()
)
```

fredrb commented on tools/local-compliance/loadGeneratingNodeScanner.go:
Also this file uses camel casing. Please change it to snake case.

fredrb commented on sensor/common/compliance/service_impl.go:
```diff
@@ -96,7 +96,11 @@ func (s *serviceImpl) startSendingLoop() {
-			con := s.connectionManager.connectionMap[msg.Hostname]
+			con, ok := s.connectionManager.connectionMap[msg.Hostname]
+			if !ok {
+				log.Errorf("unable to find connection to compliance: %q", msg.Hostname)
```
I think these logs should start with upper case
```suggestion
				log.Errorf("Unable to find connection to compliance: %q", msg.Hostname)
```

vikin91 replied:
All error logs should start with non-capitalized letter. Why do you think we should change it here? 

I see we still mix styles, but I think there was a govet rule to check this.
![Screenshot 2023-05-23 at 11 46 32](https://github.com/stackrox/stackrox/assets/114479/49041954-37a2-471a-b808-1833dce029c9)

vikin91 replied:
I tried to search for some "literature" to support this: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

fredrb replied:
> I tried to search for some "literature" to support this: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

`fmt.Errorf` and `errors.New` should definitely be lower cased, because error messages can be wrapped, and the formatting will be weird since upper case letters will appear in the middle of the error message. 

However, my comment is regarding **logging** messages. FWIW, the link you pasted states this point exactly:

> This does not apply to logging, which is implicitly line-oriented and not combined inside other messages.

I believe these examples in your screenshot are the ones not following the Go pattern. I wonder why "unable" was often capitalized (although I see some exceptions). My code search looks a bit different:

![image](https://github.com/stackrox/stackrox/assets/6811076/4d8d5621-7d12-4c94-ae66-58612b080b6d)

vikin91 replied:
Ok, you got point :) Let me change it

vikin91 replied:
I hope, I covered most of the affected code in [31b4a11](https://github.com/stackrox/stackrox/pull/5999/commits/31b4a11525ec6adc163f020e748fbefa02b0746b)

fredrb commented on compliance/collection/compliance/sensorReplyHandler.go:
```diff
@@ -0,0 +1,45 @@
+		} else {
+			err := client.Send(msg)
+			if err != nil {
+				s.log.Errorf("error sending to sensor: %v", err)
```
```suggestion
				s.log.Errorf("Error sending to sensor: %v", err)
```

fredrb commented on compliance/collection/compliance/sensorReplyHandler.go:
```diff
@@ -0,0 +1,45 @@
+		time.Sleep(time.Second * 10)
+		msg, err := s.nodeScanner.ScanNode(ctx)
+		if err != nil {
+			s.log.Errorf("error running ScanNode: %v", err)
```
```suggestion
			s.log.Errorf("Error running ScanNode: %v", err)
```

fredrb commented on compliance/collection/compliance/sensorReplyHandler.go:
```diff
@@ -0,0 +1,45 @@
+func (s *SensorReplyHandlerImpl) HandleNACK(ctx context.Context, client sensor.ComplianceService_CommunicateClient) {
+	s.log.Infof("Received NACK from Sensor, resending NodeInventory in 10 seconds.")
+	go func() {
+		time.Sleep(time.Second * 10)
```
Didn't we add a comment here about changing this into a ticker? I'll mention what we already discussed here just to not lose the comment:

You can `select` on the ticker and on a stop signal here. This is in case local-compliance has to shutdown before the 10s timer is over.

fredrb commented on compliance/collection/compliance/nodeScanner.go:
```diff
@@ -0,0 +1,83 @@
+
+// NodeInventoryComponentScanner connects to node-inventory container to provide node-inventory object
+type NodeInventoryComponentScanner struct {
+	log              *logging.Logger
```
Log as dependency here again, why?

fredrb commented on compliance/collection/compliance/nodeName.go:
This file is also in camel case, please change.

fredrb commented on compliance/collection/compliance/nodeScanner.go:
This file is also in camel case, please change.

fredrb commented on compliance/collection/compliance/sensorReplyHandler.go:
This file is also in camel case, please change.

vikin91 reviewed (commented):
Posting sync-review comments

fredrb reviewed (changes requested):
Two major things I'd like to see addressed here are:

1) The casing in the file names
2) Log module as a dependency

The rest are just nitpicks.

For (2), I don't know what all the implications of doing what you did. It's probably not too much of an issue. I could be convinced otherwise, but right now I see three reasons why we shouldn't do it like this and instead define the log as a package-wide `var`.

1. It might mess up the log message prefix. The logger is called as `LoggerForModule()` which parses the current module path to define the prefix in the message. Passing a main built logger might generate confusing logging in this case.
2. It's one more dependency to pass in the constructors (or just in the struct definition here), which could be avoided. 
3. It breaks the pattern we already have in the app. A quick search in the repo hasn't shown me a single instance where we don't instantiate the logger as a local var.

vikin91 commented on tools/local-compliance/load_generating_node_scanner.go:
```diff
@@ -0,0 +1,62 @@
+
+// GetIntervals returns an object with delay-intervals between scans
+func (n *LoadGeneratingNodeScanner) GetIntervals() *intervals.NodeScanIntervals {
+	return intervals.NewNodeScanInterval(1000*time.Millisecond, 0.0, time.Second)
```
The goal was to have a placeholder here, but I think, I could export it to an env variable. Let me add this

vikin91 replied:
Addressed in: [26f2858](https://github.com/stackrox/stackrox/pull/5999/commits/26f2858f44b8dfb44066986f9b157869916716af)

vikin91 commented:
/test ocp-4-12-core-bpf-qa-e2e-tests

vikin91 commented:
/retest

lvalerom commented on tools/local-compliance/scripts/local-compliance.sh:
```diff
@@ -0,0 +1,33 @@
+kubectl -n stackrox get secrets collector-tls -ojsonpath='{.data.collector-key\.pem}' \
+    | base64 -D -o "${DIR}/../certs/collector-key.pem"
+
+ROX_NODE_SCANNING_MAX_INITIAL_WAIT="1s" \
+ROX_NODE_SCANNING_INTERVAL="1s" \
```
I think it would be good to have these two configurable.

vikin91 replied:
You mean configurable as params of the shell script?

lvalerom replied:
Yes!

vikin91 replied:
Let me address that in a follow-up if that is fine. I would like to get this merged to unlock the work on the last phase of [ROX-15458](https://issues.redhat.com/browse/ROX-15458)

lvalerom reviewed (approved):
LGTM. I think there a couple of things that can be improved in future iterations of `local-compliance`.

1 - I would add the `runChecks` and the `AuditLogCollection` as dependencies. This will allow us to test these functionalities easily with `local-compliance`.
2 - I would pass a channel to `runChecks` and `AuditLogCollection` instead of the gRPC client. That would unify the logic that send things to sensor in `manageSendToSensor`.

vikin91 commented:
> LGTM. I think there a couple of things that can be improved in future iterations of `local-compliance`.
> 
> 1 - I would add the `runChecks` and the `AuditLogCollection` as dependencies. This will allow us to test these functionalities easily with `local-compliance`. 
> 2 - I would pass a channel to `runChecks` and `AuditLogCollection` instead of the gRPC client. That would unify the logic that send things to sensor in `manageSendToSensor`.

I started hacking on 1) already, maybe I will champlon it soon ;) The 2) is a great suggestion for further refactoring, thanks!

vikin91 commented:
@fredrb I think, you still need to unlock this PR :) Hope that I addressed your feedback. Could you give this a look at the beginning of the week please? Thanks!

fredrb reviewed (approved):
Thanks for being patient. Looks good to me now! 🚢

vikin91 commented:
I merged https://github.com/stackrox/stackrox/pull/6177 first and needed to rebase this. Will wait for CI now and merge afterwards if everything looks good here
